### PR TITLE
[BEAM-6883] Improve long-running spark streaming test.

### PR DESCRIPTION
**Pull request by RyanSkraba**
https://issues.apache.org/jira/browse/BEAM-6883

Currently, this test takes 10 minutes, and unfortunately doesn't test what is intended.  The transformation under test is an extremely long-lived BoundedSource, not an unbounded source.

I changed the transformation to an UnboundedSource, and  modified the MetricsResultMatchers to be able to apply any arbitrary Matcher to a metrics value.  This may be useful for BEAM-1169

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/StreamingSourceMetricsTest.java [hunk 1]
```diff
@@ -47,11 +52,18 @@
...
+    // Use a GenerateSequence for the UnboundedSequence, but push the watermark to infinity at
+    // minElements to let the test pipeline cleanly shut it down.  Shutdown will occur shortly
+    // afterwards, but at least minElements will be reported in the metrics.
+    PCollection<Long> pc =
+        pipeline.apply(
+            GenerateSequence.from(1)
+                .withRate(minElements / 10, Duration.millis(500L))
+                .withTimestampFn(
+                    t -> t < minElements ? Instant.now() : BoundedWindow.TIMESTAMP_MAX_VALUE));
+    assertThat(pc.isBounded(), is(PCollection.IsBounded.UNBOUNDED));
```
Comment (RyanSkraba) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/StreamingSourceMetricsTest.java]: Note: this assert would have failed with the old GenerateSequence configuration.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java [hunk 2]
```diff
@@ -43,9 +43,50 @@
...
+  }
+
   /**
    * Matches a {@link MetricResult} with the given namespace, name and step, and whose value equals
    * the given value for either committed or attempted (based on {@code isCommitted}) metrics.
+   *
+   * <p>For metrics with a {@link {@link GaugeResult}}, only the value is matched and the timestamp
+   * is ignored.
    */
   public static <T> Matcher<MetricResult<T>> metricsResult(
```
Comment (RyanSkraba) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java]: Another note: I checked that these modifications to existing methods produce the same error messages as before by manually forcing other existing metrics tests to fail.

**Comment by RyanSkraba** [comment]
Run Spark ValidatesRunner

**Comment by RyanSkraba** [comment]
R: @dmvk
CC: @iemejia, @echauchot

**Comment by RyanSkraba** [comment]
The links to compare:

* Master (10 minutes): https://builds.apache.org/view/A-D/view/Beam/view/PostCommit/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastBuild/testReport/org.apache.beam.runners.spark.translation.streaming/

* PR (20 seconds):  https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_PR/112/testReport/org.apache.beam.runners.spark.translation.streaming/

**Comment by dmvk** [comment]
Run Direct ValidatesRunner

File: sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java (hunk 2 above)
Comment (dmvk) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java]: If I understand this correctly, you did diff of failing tests output before and after the change and there was no difference?

Comment (RyanSkraba) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java]: Exactly: I modified some tests to fail and checked the output before and after the change.

*For exact match:*

```
Expected: a collection containing MetricResult{inNamespace="source", name="elements_read", step="GenerateSequence/Read(UnboundedCountingSource)", attempted=<999L>}
     but: mismatches were: [MetricResult{attempted: <999L> != <1000L>}]
```

*Using a matcher (new):*

```
Expected: a collection containing MetricResult{inNamespace="source", name="elements_read", step="GenerateSequence/Read(UnboundedCountingSource)", attempted=a value equal to or greater than <1000000L>}
     but: mismatches were: [MetricResult{attempted: <1000L> was less than <1000000L>}]
```

**Comment by RyanSkraba** [comment]
Hello!  Just to check -- I did a rebase to change this into two commits so they can be merged without squashing if you prefer...  I'm a bit unclear whether `!fixup` commits work from the github UI, and if review history is lost!

**Comment by dmvk** [comment]
Run Java_Examples_Dataflow PreCommit

**Pull request merged**